File: vm/picostore.js
// picostore.js -- PicoStore (packs + CRUD) and the card query language, in JS.
// Mirrors picostore.py; result-identical queries and byte-identical card encoding
// (via picoserializer.js). A pluggable backend exposes get/set/remove/keys; the
// default is in-memory, but localStorage works directly (see the site).
(function (root, factory) {
  var SER = (typeof module !== "undefined" && module.exports) ? require("./picoserializer.js") : root.PicoSerializer;
  var P = factory(SER);
  if (typeof module !== "undefined" && module.exports) module.exports = P;
  else root.PicoStore = P;
})(typeof globalThis !== "undefined" ? globalThis : this, function (SER) {
  "use strict";

  function MemBackend() { this.d = {}; }
  MemBackend.prototype = {
    get: function (k) { return (k in this.d) ? this.d[k] : null; },
    set: function (k, v) { this.d[k] = v; },
    remove: function (k) { delete this.d[k]; },
    keys: function () { return Object.keys(this.d); }
  };

  // ---- query language ----
  var CMP2 = { "==": 1, "!=": 1, "<=": 1, ">=": 1, "<>": 1 };
  function qTokens(q) {
    var toks = [], i = 0, n = q.length;
    while (i < n) {
      var c = q[i];
      if (/\s/.test(c)) { i++; continue; }
      if (c === '"' || c === "'") { var j = i + 1, b = ""; while (j < n && q[j] !== c) { b += q[j]; j++; } toks.push(["str", b]); i = j + 1; continue; }
      var two = q.substr(i, 2);
      if (CMP2[two]) { toks.push(["op", two]); i += 2; continue; }
      if ("<>=~".indexOf(c) >= 0) { toks.push(["op", c]); i++; continue; }
      if (c === "(" || c === ")") { toks.push(["paren", c]); i++; continue; }
      var k = i; while (k < n && !/\s/.test(q[k]) && "<>=~()\"'".indexOf(q[k]) < 0) k++;
      var w = q.slice(i, k); i = k;
      var up = w.toUpperCase();
      if (up === "AND" || up === "OR") toks.push(["kw", up]); else toks.push(["word", w]);
    }
    return toks;
  }
  function coerce(tok) {
    if (tok[0] === "str") return tok[1];
    var t = tok[1];
    if (/^-?\d+$/.test(t)) return parseInt(t, 10);
    if (/^0[xX][0-9a-fA-F]+$/.test(t)) return parseInt(t, 16);
    return t;
  }
  function QParser(toks) { this.toks = toks; this.i = 0; }
  QParser.prototype = {
    peek: function () { return this.i < this.toks.length ? this.toks[this.i] : null; },
    nxt: function () { return this.toks[this.i++]; },
    parse: function () { return this.parseOr(); },
    parseOr: function () { var left = this.parseAnd(); var p; while ((p = this.peek()) && p[0] === "kw" && p[1] === "OR") { this.nxt(); left = ["or", left, this.parseAnd()]; } return left; },
    parseAnd: function () { var left = this.parseCmp(); var p; while ((p = this.peek()) && p[0] === "kw" && p[1] === "AND") { this.nxt(); left = ["and", left, this.parseCmp()]; } return left; },
    parseCmp: function () {
      var p = this.peek();
      if (p && p[0] === "paren" && p[1] === "(") { this.nxt(); var node = this.parseOr(); var q = this.peek(); if (q && q[0] === "paren" && q[1] === ")") this.nxt(); return node; }
      var field = this.nxt(); if (!field || field[0] !== "word") throw new Error("query: expected field");
      var op = this.nxt(); if (!op || op[0] !== "op") throw new Error("query: expected operator");
      var val = this.nxt(); if (!val) throw new Error("query: expected value");
      return ["cmp", field[1], op[1], coerce(val)];
    }
  };
  function evalCmp(field, op, value, rec) {
    if (!(field in rec)) return false;
    var fv = rec[field];
    if (op === "=" || op === "==") return fv === value;
    if (op === "!=" || op === "<>") return fv !== value;
    if (op === "~") return String(fv).indexOf(String(value)) >= 0;
    if (typeof fv !== typeof value) return false;
    if (op === "<") return fv < value;
    if (op === ">") return fv > value;
    if (op === "<=") return fv <= value;
    if (op === ">=") return fv >= value;
    throw new Error("query: unknown operator " + op);
  }
  function evalNode(node, rec) {
    if (node[0] === "and") return evalNode(node[1], rec) && evalNode(node[2], rec);
    if (node[0] === "or") return evalNode(node[1], rec) || evalNode(node[2], rec);
    return evalCmp(node[1], node[2], node[3], rec);
  }
  function compileQuery(q) {
    q = (q || "").trim();
    if (!q) return function () { return true; };
    var ast = new QParser(qTokens(q)).parse();
    return function (rec) { return evalNode(ast, rec); };
  }

  // ---- store ----
  function PicoStore(backend) { this.b = backend || new MemBackend(); }
  PicoStore.prototype = {
    _ids: function (pack) { var raw = this.b.get(pack + ":ids"); return raw ? raw.split(",").filter(Boolean).map(Number) : []; },
    _setIds: function (pack, ids) { this.b.set(pack + ":ids", ids.join(",")); },
    create: function (pack, record) {
      var nxt = parseInt(this.b.get(pack + ":next") || "1", 10);
      this.b.set(pack + ":card:" + nxt, SER.toHex(SER.serializeCard(record)));
      this._setIds(pack, this._ids(pack).concat([nxt]));
      this.b.set(pack + ":next", String(nxt + 1));
      return nxt;
    },
    read: function (pack, id) { var h = this.b.get(pack + ":card:" + id); return h ? SER.deserializeCard(SER.fromHex(h)) : null; },
    update: function (pack, id, record) { if (this._ids(pack).indexOf(id) < 0) return false; this.b.set(pack + ":card:" + id, SER.toHex(SER.serializeCard(record))); return true; },
    patch: function (pack, id, fields) { var r = this.read(pack, id); if (!r) return false; for (var k in fields) r[k] = fields[k]; return this.update(pack, id, r); },
    delete: function (pack, id) { var ids = this._ids(pack); var idx = ids.indexOf(id); if (idx < 0) return false; ids.splice(idx, 1); this._setIds(pack, ids); this.b.remove(pack + ":card:" + id); return true; },
    all: function (pack) { var self = this; return this._ids(pack).map(function (id) { return [id, self.read(pack, id)]; }).filter(function (e) { return e[1] !== null; }); },
    query: function (pack, q) { var pred = compileQuery(q); return this.all(pack).filter(function (e) { return pred(e[1]); }); },
    cardBytesHex: function (pack, id) { return this.b.get(pack + ":card:" + id); }
  };

  return { PicoStore: PicoStore, MemBackend: MemBackend, compileQuery: compileQuery };
});

File: picostore.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""picostore.py -- PicoStore: a pack-based card store with CRUD and a small query
language, backed by PicoBinarySerializer (picoserializer.py).

A *pack* is a named collection of cards (records) keyed by an auto-incrementing
id. Cards are serialized to binary and held in a pluggable key/value byte backend
(an in-memory dict here; localStorage in the browser, see vm/picostore.js).

CRUD:
    sid = store.create(pack, {"qty": 42, "sku": "ABC"})
    rec = store.read(pack, sid)
    store.update(pack, sid, {...})
    store.delete(pack, sid)

Query language (string):
    qty > 40 AND sku ~ "AB"
    status = 1 OR qty <= 0
Operators: =  ==  !=  <>  <  >  <=  >=  ~ (string contains).  AND / OR combine
comparisons; AND binds tighter than OR. Field values are int literals, quoted
strings, or barewords. A missing field never matches.
"""

from __future__ import annotations

from typing import Callable, Dict, List, Optional, Tuple

from picoserializer import serialize_card, deserialize_card, to_hex, from_hex


# ── key/value byte backend ───────────────────────────────────────────────────

class DictBackend:
    """Default in-memory backend. Values are strings (hex / csv / ints)."""

    def __init__(self):
        self._d: Dict[str, str] = {}

    def get(self, key: str) -> Optional[str]:
        return self._d.get(key)

    def set(self, key: str, value: str) -> None:
        self._d[key] = value

    def remove(self, key: str) -> None:
        self._d.pop(key, None)

    def keys(self) -> List[str]:
        return list(self._d.keys())


# ── query language ───────────────────────────────────────────────────────────

_CMP2 = {"==", "!=", "<=", ">=", "<>"}


def _q_tokens(q: str) -> List[Tuple[str, str]]:
    toks: List[Tuple[str, str]] = []
    i, n = 0, len(q)
    while i < n:
        c = q[i]
        if c.isspace():
            i += 1; continue
        if c in "\"'":
            j = i + 1; buf = []
            while j < n and q[j] != c:
                buf.append(q[j]); j += 1
            toks.append(("str", "".join(buf))); i = j + 1; continue
        two = q[i:i + 2]
        if two in _CMP2:
            toks.append(("op", two)); i += 2; continue
        if c in "<>=~":
            toks.append(("op", c)); i += 1; continue
        if c in "()":
            toks.append(("paren", c)); i += 1; continue
        j = i
        while j < n and (not q[j].isspace()) and q[j] not in "<>=~()\"'":
            j += 1
        w = q[i:j]; i = j
        up = w.upper()
        if up in ("AND", "OR"):
            toks.append(("kw", up))
        else:
            toks.append(("word", w))
    return toks


class _QParser:
    def __init__(self, toks):
        self.toks = toks; self.i = 0

    def peek(self):
        return self.toks[self.i] if self.i < len(self.toks) else None

    def nxt(self):
        t = self.toks[self.i]; self.i += 1; return t

    def parse(self):
        node = self.parse_or()
        return node

    def parse_or(self):
        left = self.parse_and()
        while self.peek() and self.peek() == ("kw", "OR"):
            self.nxt(); right = self.parse_and()
            left = ("or", left, right)
        return left

    def parse_and(self):
        left = self.parse_cmp()
        while self.peek() and self.peek() == ("kw", "AND"):
            self.nxt(); right = self.parse_cmp()
            left = ("and", left, right)
        return left

    def parse_cmp(self):
        if self.peek() and self.peek()[0] == "paren" and self.peek()[1] == "(":
            self.nxt(); node = self.parse_or()
            if self.peek() and self.peek() == ("paren", ")"):
                self.nxt()
            return node
        field = self.nxt()
        if field[0] != "word":
            raise ValueError(f"query: expected field, got {field}")
        op = self.nxt()
        if op[0] != "op":
            raise ValueError(f"query: expected operator, got {op}")
        val = self.nxt()
        return ("cmp", field[1], op[1], _coerce(val))


def _coerce(tok):
    kind, text = tok
    if kind == "str":
        return text
    try:
        return int(text, 0)
    except (ValueError, TypeError):
        return text


def _eval_cmp(field, op, value, rec):
    if field not in rec:
        return False
    fv = rec[field]
    if op in ("=", "=="):
        return fv == value
    if op in ("!=", "<>"):
        return fv != value
    if op == "~":
        return str(value) in str(fv)
    # ordered comparisons
    try:
        if op == "<":
            return fv < value
        if op == ">":
            return fv > value
        if op == "<=":
            return fv <= value
        if op == ">=":
            return fv >= value
    except TypeError:
        return False
    raise ValueError(f"query: unknown operator {op}")


def _eval(node, rec) -> bool:
    kind = node[0]
    if kind == "and":
        return _eval(node[1], rec) and _eval(node[2], rec)
    if kind == "or":
        return _eval(node[1], rec) or _eval(node[2], rec)
    return _eval_cmp(node[1], node[2], node[3], rec)


def compile_query(q: str) -> Callable[[dict], bool]:
    """Compile a query string to a predicate `record -> bool`. Empty = match all."""
    q = (q or "").strip()
    if not q:
        return lambda rec: True
    ast = _QParser(_q_tokens(q)).parse()
    return lambda rec: _eval(ast, rec)


# ── store ────────────────────────────────────────────────────────────────────

class PicoStore:
    def __init__(self, backend=None):
        self.b = backend if backend is not None else DictBackend()

    def _ids(self, pack) -> List[int]:
        raw = self.b.get(f"{pack}:ids")
        return [int(x) for x in raw.split(",") if x] if raw else []

    def _set_ids(self, pack, ids):
        self.b.set(f"{pack}:ids", ",".join(str(x) for x in ids))

    def create(self, pack: str, record: dict) -> int:
        nxt = int(self.b.get(f"{pack}:next") or "1")
        self.b.set(f"{pack}:card:{nxt}", to_hex(serialize_card(record)))
        self._set_ids(pack, self._ids(pack) + [nxt])
        self.b.set(f"{pack}:next", str(nxt + 1))
        return nxt

    def read(self, pack: str, card_id: int) -> Optional[dict]:
        hexs = self.b.get(f"{pack}:card:{card_id}")
        return deserialize_card(from_hex(hexs)) if hexs else None

    def update(self, pack: str, card_id: int, record: dict) -> bool:
        if card_id not in self._ids(pack):
            return False
        self.b.set(f"{pack}:card:{card_id}", to_hex(serialize_card(record)))
        return True

    def patch(self, pack: str, card_id: int, fields: dict) -> bool:
        rec = self.read(pack, card_id)
        if rec is None:
            return False
        rec.update(fields)
        return self.update(pack, card_id, rec)

    def delete(self, pack: str, card_id: int) -> bool:
        ids = self._ids(pack)
        if card_id not in ids:
            return False
        ids.remove(card_id)
        self._set_ids(pack, ids)
        self.b.remove(f"{pack}:card:{card_id}")
        return True

    def all(self, pack: str) -> List[Tuple[int, dict]]:
        out = []
        for cid in self._ids(pack):
            rec = self.read(pack, cid)
            if rec is not None:
                out.append((cid, rec))
        return out

    def query(self, pack: str, q: str) -> List[Tuple[int, dict]]:
        pred = compile_query(q)
        return [(cid, rec) for cid, rec in self.all(pack) if pred(rec)]

    def card_bytes_hex(self, pack: str, card_id: int) -> Optional[str]:
        return self.b.get(f"{pack}:card:{card_id}")
